### PR TITLE
IO-172/171/168-Project-CRUD

### PR DESCRIPTION
**Pull request by Git-Uzair**
Implements CRUD functionality for all the Project and related information.
Implemented new Models, better serialization of data.
Written tests for Projects, Task and Budget item.
Solved migration issues

Overlaps with IO-172,171 & 168. All these tasks can be reviewed through this PR

### CODE REVIEW

**Review by kgunga — CHANGES_REQUESTED** [review]
We need also initial static data for dropdown menus like ProjectPhase, ProjectType. Please check other dropdown menus that need static data. Please populate database in a manner that description is in Finnish but its key is in english.

File: infraohjelmointi_api/models/ProjectSet.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    responsiblePerson = models.ForeignKey(
+        "Person", on_delete=models.DO_NOTHING, blank=True, null=True
+    )
+    phase = models.CharField(
+        max_length=16,
+        choices=ProjectPhaseChoices.choices,
+        default=ProjectPhaseChoices.PROPOSAL,
+    )
+    programmed = models.BooleanField(default=False)
+    # finances = models.TextField(max_length=500, blank=True, null=True)
```
[the remark this answers is not in the document]
Comment (Git-Uzair) [reply infraohjelmointi_api/models/ProjectSet.py:32]: Yes, not sure if this field is Text or integer

File: infraohjelmointi_api/models/ProjectSet.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    # def sapProjects(self):
```
[the remark this answers is not in the document]
Comment (Git-Uzair) [reply infraohjelmointi_api/models/ProjectSet.py]: Removed

File: infraohjelmointi_api/tests/test_Tasks.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
+from django.test import TestCase
+from rest_framework.renderers import JSONRenderer
+from infraohjelmointi_api.models import Person, Project, ProjectType
+from infraohjelmointi_api.serializers import TaskSerializer
+from ..models import Task
+import uuid
+
+
+class TaskTestCase(TestCase):
+    TaskId = uuid.uuid4()
```
[the remark this answers is not in the document]
Comment (Git-Uzair) [reply infraohjelmointi_api/tests/test_Tasks.py]: Created static UUID

File: infraohjelmointi_api/tests/test_Tasks.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+    @classmethod
+    def setUpTestData(self):
+
+        self.projectType = ProjectType.objects.create(
+            id=uuid.uuid4(), value="projectComplex"
```
[the remark this answers is not in the document]
Comment (Git-Uzair) [reply infraohjelmointi_api/tests/test_Tasks.py]: Created static UUID

File: infraohjelmointi_api/tests/test_Tasks.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        )
+        self.person_1 = Person.objects.create(
+            id=uuid.uuid4(),
```
[the remark this answers is not in the document]
Comment (Git-Uzair) [reply infraohjelmointi_api/tests/test_Tasks.py]: Created static UUID

File: infraohjelmointi_api/tests/test_Tasks.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+            firstName="John",
+            lastName="Doe",
+            email="random@random.com",
+            title="Manager",
+            phone="0414853275",
+        )
+
+        self.project = Project.objects.create(
+            id=uuid.uuid4(),
```
[the remark this answers is not in the document]
Comment (Git-Uzair) [reply infraohjelmointi_api/tests/test_Tasks.py]: Created static UUID

File: infraohjelmointi_api/tests/test_Tasks.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+            hkrId=43210,
+            sapProject=uuid.uuid4(),
```
[the remark this answers is not in the document]
Comment (Git-Uzair) [reply infraohjelmointi_api/tests/test_Tasks.py]: Created static UUID

File: infraohjelmointi_api/models/ProjectSet.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
+import uuid
+from django.db import models
+from django.utils.translation import gettext_lazy as lazy
+
+
+class ProjectSet(models.Model):
+    class ProjectPhaseChoices(models.TextChoices):
```
[the remark this answers is not in the document]
Comment (Git-Uzair) [reply infraohjelmointi_api/models/ProjectSet.py]: Made it a foreign key

File: infraohjelmointi_api/tests/test_Projects.py [hunk 9]
```diff
@@ -90,8 +90,8 @@ def setUpTestData(self):
             id=self.projectId,
             siteId=self.budgetItem,
             hkrId=12345,
-            sapProject=uuid.uuid4(),
-            sapNetwork=uuid.uuid4(),
+            sapProject=[uuid.uuid4().__str__()],
```
Comment (kgunga) [inline infraohjelmointi_api/tests/test_Projects.py:93]: static uuids, please

File: infraohjelmointi_api/tests/test_Projects.py [hunk 10]
```diff
@@ -239,8 +239,8 @@ def test_GET_all_projects(self):
             id=self.projectId2,
             siteId=self.budgetItem,
             hkrId=2265,
-            sapProject=uuid.uuid4(),
-            sapNetwork=uuid.uuid4(),
+            sapProject=[uuid.uuid4().__str__()],
```
Comment (kgunga) [inline infraohjelmointi_api/tests/test_Projects.py:242]: static uuids, please

File: infraohjelmointi_api/tests/test_Tasks.py [hunk 11]
```diff
@@ -37,8 +37,8 @@ def setUpTestData(self):
         self.project = Project.objects.create(
             id=self.projectId,
             hkrId=43210,
-            sapProject=self.sapProjectId,
-            sapNetwork=self.sapNetworkId,
+            sapProject=[uuid.uuid4().__str__()],
```
Comment (kgunga) [inline infraohjelmointi_api/tests/test_Tasks.py:40]: static uuids, please

File: infraohjelmointi_api/tests/test_Tasks.py [hunk 12]
```diff
@@ -37,8 +37,8 @@ def setUpTestData(self):
...
+            sapNetwork=[uuid.uuid4().__str__()],
```
Comment (kgunga) [inline infraohjelmointi_api/tests/test_Tasks.py:41]: static uuids, please

File: infraohjelmointi_api/serializers.py [hunk 13]
```diff
@@ -25,54 +25,30 @@ class Meta:
...
         model = Person
 
         exclude = ["createdDate", "updatedDate"]
 
 
-class ProjectTypeValOnlySerializer(serializers.ModelSerializer):
-    def to_representation(self, instance):
-        """Convert returns only String instead of Object of value"""
-        ret = super().to_representation(instance)
+# class ProjectTypeValOnlySerializer(serializers.ModelSerializer):
```
Comment (kgunga) [inline infraohjelmointi_api/serializers.py:42]: remove commented code

**Review by kgunga — CHANGES_REQUESTED** [review]
Please remove all commented code and use static uuids in tests

**Review by kgunga — APPROVED** [review]
👍

**Pull request merged**